From 4fb4fa69ccda274702364fe01fedf985abdeb8fc Mon Sep 17 00:00:00 2001
From: Xinyu Chen <xinyu.chen@freescale.com>
Date: Tue, 7 Aug 2012 10:28:14 +0800
Subject: [PATCH 323/635] ENGR00219578 ion: refine the ion get physical
 address ioctl

The original design will return failed when address is
higher than 0x80000000.
ioctl should only return a result, but put address
into the param structures.

Signed-off-by: Xinyu Chen <xinyu.chen@freescale.com>
---
 ion/ion.c | 12 +++++++++---
 ion/ion.h |  2 +-
 2 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/ion/ion.c b/ion/ion.c
index c0f1873..480add0 100644
--- a/ion/ion.c
+++ b/ion/ion.c
@@ -127,12 +127,18 @@ int ion_map(int fd, struct ion_handle *handle, size_t length, int prot,
         return ret;
 }
 
-int ion_phys(int fd, struct ion_handle *handle)
+unsigned long ion_phys(int fd, struct ion_handle *handle)
 {
-        struct ion_handle_data data = {
+        int ret;
+        struct ion_phys_data data = {
                 .handle = handle,
+		.phys = 0,
         };
-        return ion_ioctl(fd, ION_IOC_PHYS, &data);
+
+        ret = ion_ioctl(fd, ION_IOC_PHYS, &data);
+        if (ret == 0)
+            return data.phys;
+        return 0;
 }
 
 int ion_share(int fd, struct ion_handle *handle, int *share_fd)
diff --git a/ion/ion.h b/ion/ion.h
index b316e61..5d77751 100644
--- a/ion/ion.h
+++ b/ion/ion.h
@@ -33,5 +33,5 @@ int ion_map(int fd, struct ion_handle *handle, size_t length, int prot,
             int flags, off_t offset, unsigned char **ptr, int *map_fd);
 int ion_share(int fd, struct ion_handle *handle, int *share_fd);
 int ion_import(int fd, int share_fd, struct ion_handle **handle);
-int ion_phys(int fd, struct ion_handle *handle);
+unsigned long ion_phys(int fd, struct ion_handle *handle);
 
-- 
1.8.0

